Questions about building the project

At first I clone the whole project, following the guide in the readme. My system is MacOS BigSur. Running 


> mkdir build

> cmake ..

> make -j8

There is a long error, the key part is 

ld: warning: ignoring file ../embree/lib-linux/libembree3.so, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
ld: warning: ignoring file ../embree/lib-linux/libembree3.so, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
Undefined symbols for architecture x86_64:
  "_rtcAttachGeometry", referenced from:
      register_embree_op::operator()(Sphere const&) const in liblajolla_lib.a(shape.cpp.o)
      register_embree_op::operator()(TriangleMesh const&) const in liblajolla_lib.a(shape.cpp.o)
      decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&) in liblajolla_lib.a(shape.cpp.o)

I thought it was the problem that the libembree.dylib is not included in the project, so I modified the original cmakelist.txt.
[CMakeLists.txt](https://github.com/BachiLi/lajolla_public/files/7805124/CMakeLists.txt)
With this cmakelists, I can build the project successfully without running the tests on intersection. However, that is where embree makes effect. The error log is the same as above.  

ld: warning: ignoring file ../embree/lib-linux/libembree3.so, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
ld: warning: ignoring file ../embree/lib-linux/libembree3.so, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
Undefined symbols for architecture x86_64:
  "_rtcAttachGeometry", referenced from:
      register_embree_op::operator()(Sphere const&) const in liblajolla_lib.a(shape.cpp.o)
      register_embree_op::operator()(TriangleMesh const&) const in liblajolla_lib.a(shape.cpp.o)
      decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&) in liblajolla_lib.a(shape.cpp.o)

I thought it may be caused by search order. Following the guide in the discord, I swap the order between lib-linux and lib-macos in the FindEmbree.cmake, but nothing happens. I now even wonder what the real problem is.

/usr/local/Cellar/cmake/3.22.1/bin/cmake -E cmake_link_script CMakeFiles/lajolla.dir/link.txt --verbose=1
/Library/Developer/CommandLineTools/usr/bin/c++ -O2 -g -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/lajolla.dir/src/main.cpp.o -o lajolla  liblajolla_lib.a 
Undefined symbols for architecture x86_64:
  "_rtcAttachGeometry", referenced from:
      register_embree_op::operator()(Sphere const&) const in liblajolla_lib.a(shape.cpp.o)
      register_embree_op::operator()(TriangleMesh const&) const in liblajolla_lib.a(shape.cpp.o)
      decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&) in liblajolla_lib.a(shape.cpp.o)
  "_rtcCommitGeometry", referenced from:
      register_embree_op::operator()(Sphere const&) const in liblajolla_lib.a(shape.cpp.o)
      register_embree_op::operator()(TriangleMesh const&) const in liblajolla_lib.a(shape.cpp.o)
      decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<register_embree_op>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Sphere, TriangleMesh> const&) in liblajolla_lib.a(shape.cpp.o)

--- a/cmake/FindEmbree.cmake
+++ b/cmake/FindEmbree.cmake
@@ -4,13 +4,22 @@ find_path(EMBREE_INCLUDE_PATH embree3/rtcore.h
   /usr/local/include
   /opt/local/include)
 
-find_library(EMBREE_LIBRARY NAMES embree3 PATHS 
-  ${CMAKE_SOURCE_DIR}/embree/lib-linux
+if (APPLE)
+find_library(EMBREE_LIBRARY NAMES embree3 PATHS
   ${CMAKE_SOURCE_DIR}/embree/lib-macos
-  ${CMAKE_SOURCE_DIR}/embree/lib-win32
-  /usr/lib 
-  /usr/local/lib 
+  /usr/lib
+  /usr/local/lib
   /opt/local/lib)
+elseif (WIN32)
+find_library(EMBREE_LIBRARY NAMES embree3 PATHS
+  ${CMAKE_SOURCE_DIR}/embree/lib-win32)
+else ()
+find_library(EMBREE_LIBRARY NAMES embree3 PATHS
+  ${CMAKE_SOURCE_DIR}/embree/lib-linux
+  /usr/lib
+  /usr/local/lib
+  /opt/local/lib)
+endif ()
 
 if (EMBREE_INCLUDE_PATH AND EMBREE_LIBRARY)
   set(EMBREE_FOUND TRUE)

